src/eve/assets.ts

```diff
@@ -0,0 +1,170 @@
+import {
+  ESI_CHARACTERS_$characterId_ASSETS,
+  ESI_CHARACTERS_$characterId_ASSETS_NAMES,
+} from '../data-source/esi/endpoints';
+import { EsiAsset } from '../data-source/esi/EsiAsset';
+import { fetchEsi, fetchEsiEx } from '../data-source/esi/fetch/fetchEsi';
+import { dao } from '../db/dao';
+import { SdeType } from '../db/tables';
+import { Tnex } from '../db/tnex';
+import { arrayToMap } from '../util/collections';
+import { TYPE_CATEGORY_SHIP } from './constants/categories';
+
+export type AssetLocationType = 'station' | 'solar_system' | 'item' | 'other';
+
+export interface Asset {
+  // Unique ID of this item.
+  itemId: number;
+  // Name, if this is an assembled ship. There are other named items in EVE,
+  // but we currently don't bother fetching their names.
+  name?: string;
+  // Type of the item.
+  typeId: number;
+  typeCategory: number;
+  typeName: string;
+  // Is it stackable?
+  isSingleton: boolean;
+  // How many in this stack? 1 if not stackable.
+  quantity: number;
+  // ID of the location.
+  locationId: number;
+  // Type of the location; note that "item" may refer to an Upwell structure
+  // or to a container item like a ship.
+  locationType: AssetLocationType;
+  // Where within a container the item is located; e.g. "Hangar" or "HiSlot3".
+  locationFlag: string;
+  // If this is a blueprint, is it a copy?
+  isBlueprintCopy?: boolean;
+}
+
+/**
+ * Reformats asset location flag to a human-readable form. E.g. "FleetHangar"
+ * becomes "fleet hangar".
+ */
+export function formatLocationFlag(locationFlag: string): string {
+  return locationFlag
+    .replace(/([A-Z0-9])/g, ' $1')
+    .trim()
+    .toLowerCase();
+}
+
+/** Retrieves assets for a given character from ESI. */
+export async function fetchAssets(
+  characterId: number,
+  token: string,
+  db: Tnex,
+): Promise<Asset[]> {
+  let assets: EsiAsset[] = [];
+  for (let page = 1; true; page++) {
```

Comment:
Would it make sense to have some sort of safety boundary here to avoid completely jamming the roster server if we get back some weird page value from ESI?

Reply:
Done